File: Programmers/[81303] 표 편집/JFe/81303.py
def solution(n, k, cmd):
    answer = ['O' for _ in range(n)]
    current, removed = k, []
    # 연결리스트 형태의 dictionary 생성 (첫 행과 마지막 행 연결 부분은 None)
    table = {i: [i-1, i+1] for i in range(1, n-1)}
    table[0] = [None, 1]
    table[n-1] = [n-2, None]
    for c in cmd:
        if c[0] == 'D':
            for _ in range(int(c[2:])):
                current = table[current][1]
        elif c[0] == 'U':
            for _ in range(int(c[2:])):
                current = table[current][0]
        elif c[0] == 'C':
            answer[current] = 'X'
            prev, next = table[current]
            removed.append((table[current], current))   # 삭제된 값과 위치 removed에 append
            # 첫 행, 마지막 행 삭제하는 경우 나눠서 처리
            if prev == None:
                table[next][0] = None
                current = next
            elif next == None:
                table[prev][1] = None
                current = prev
            else:
                table[prev][1] = next
                table[next][0] = prev
                current = next
        elif c[0] == 'Z':
            (prev, next), i = removed.pop()
            answer[i] = 'O'
            # 첫 행, 마지막 행 복구하는 경우 나눠서 처리
            if prev == None:
                table[next][0] = i
            elif next == None:
                table[prev][1] = i
            else:
                table[prev][1] = i
                table[next][0] = i
    return ''.join(answer)

File: Programmers/[81303] 표 편집/KingDonggyu/81303.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}

function solution(n, k, cmd) {
  const answer = [];
  const removed = [];

  let selected = null;
  let head = null;

  // Linked List 생성
  for (let row = 0; row < n; row++) {
    answer[row] = 'O';
    const node = new Node(row);

    if (!head) {
      head = node;
      continue;
    }

    head.next = node;
    node.prev = head;
    head = head.next;

    if (row === k) {
      selected = node;
    }
  }

  for (const s of cmd) {
    let [ch, count] = s.split(' ');
    count = Number(count);

    switch (ch) {
      case 'C':
        removed.push(selected);
        answer[selected.value] = 'X';

        if (selected.prev) {
          selected.prev.next = selected.next;
        }

        if (selected.next) {
          selected.next.prev = selected.prev;
          // 삭제 후 아래 행 선택
          selected = selected.next;
        } else {
          // 삭제 후 윗 행 선택
          selected = selected.prev;
        }
        break;
      case 'Z': 
        const restore = removed.pop();
        answer[restore.value] = 'O';

        if (restore.prev) {
          restore.prev.next = restore;
        }

        if (restore.next) {
          restore.next.prev = restore;
        }
        break;
      case 'D': 
        while (count--) {
          selected = selected.next;
        }
        break;
      case 'U': 
        while (count--) {
          selected = selected.prev;
        }
    }
  }

  return answer.join('');
}
